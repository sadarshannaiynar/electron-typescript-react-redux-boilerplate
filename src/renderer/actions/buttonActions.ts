import { IpcMessageEvent, IpcRenderer } from 'electron';
import { Dispatch } from 'redux';
import { action } from 'typesafe-actions';
import { EXEC_LS } from '../../common/constants/commandTypes';
import { DECREMENT_VALUE, EXEC_LS_ERROR, EXEC_LS_SUCCESS, INCREMENT_VALUE, SET_VALUE } from '../constants/actionTypes';

export interface IEventOutput {
  log: string;
}

export const incrementAction = () => action(INCREMENT_VALUE);

export const decrementAction = () => action(DECREMENT_VALUE);

export const setValue = (value: number) => action(SET_VALUE, value);

export const startExecuteLs = () => action(EXEC_LS);

export const successExecuteLs = (output: string) => action(EXEC_LS_SUCCESS, output);

export const errorExecuteLs = (error: string) => action(EXEC_LS_ERROR, error);

export const executeLs = (ipcRenderer: IpcRenderer): any => (
  (dispatch: Dispatch) => {
    ipcRenderer.send(EXEC_LS);
    dispatch(startExecuteLs());

    ipcRenderer.once(`success-${EXEC_LS}`, (event: IpcMessageEvent, { log }: IEventOutput) => {
      dispatch(successExecuteLs(log));
      ipcRenderer.removeAllListeners(`error-${EXEC_LS}`);
    });

    ipcRenderer.once(`error-${EXEC_LS}`, (event: IpcMessageEvent, { log }: IEventOutput) => {
      dispatch(errorExecuteLs(log));
      ipcRenderer.removeAllListeners(`success-${EXEC_LS}`);
    });
  }
);
