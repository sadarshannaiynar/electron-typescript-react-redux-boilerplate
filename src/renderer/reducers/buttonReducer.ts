import { ActionType, getType } from 'typesafe-actions';
import { EXEC_LS } from '../../common/constants/commandTypes';
import * as buttonActions from '../actions/buttonActions';
import { DECREMENT_VALUE, EXEC_LS_ERROR, EXEC_LS_SUCCESS, INCREMENT_VALUE, SET_VALUE } from '../constants/actionTypes';

export type ButtonActions = ActionType<typeof buttonActions>;

export interface IButtonReducerState {
  readonly value: number;
  readonly output: string;
  readonly error: string;
}

const defaultState: IButtonReducerState = {
  error: '',
  output: '',
  value: 0,
};

export const reducer = (state: IButtonReducerState = defaultState, action?: ButtonActions): IButtonReducerState => {
  if (action) {
    switch (action.type) {
      case INCREMENT_VALUE:
        return {
          ...state,
          value: state.value + 1,
        };
      case DECREMENT_VALUE:
        return {
          ...state,
          value: state.value - 1,
        };
      case SET_VALUE:
        return {
          ...state,
          value: action.payload,
        };
      case EXEC_LS:
        return {
          ...state,
          error: '',
          output: '',
        };
      case EXEC_LS_SUCCESS:
        return {
          ...state,
          output: action.payload,
        };
      case EXEC_LS_ERROR:
        return {
          ...state,
          error: action.payload,
        };
    }
  }
  return { ...state };
};
