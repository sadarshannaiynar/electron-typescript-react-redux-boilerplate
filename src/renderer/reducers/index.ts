import { combineReducers } from 'redux';
import * as buttonReducer from './buttonReducer';

export interface IRootState {
  buttonReducer: buttonReducer.IButtonReducerState;
}

export type RootAction =
  | buttonReducer.ButtonActions;

export const combinedReducers = combineReducers<IRootState, RootAction>({
  buttonReducer: buttonReducer.reducer,
});
