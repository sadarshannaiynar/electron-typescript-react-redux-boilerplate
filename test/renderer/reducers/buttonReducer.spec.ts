import { decrementAction, incrementAction } from '../../../src/renderer/actions/buttonActions';
import { ButtonActions, IButtonReducerState, reducer } from '../../../src/renderer/reducers/buttonReducer';

const getCurrentState = (value: number, output: string, error: string): IButtonReducerState => ({
  error,
  output,
  value,
});

describe('buttonReducer Tests', () => {
  test('should return the default state', () => {
    const nextState = reducer(undefined);
    expect(nextState).toMatchSnapshot();
    expect(nextState.value).toStrictEqual(0);
    expect(nextState.output).toStrictEqual('');
    expect(nextState.error).toStrictEqual('');
  });

  test('should use the default state when state is undefined', () => {
    const action: ButtonActions = { type: 'INCREMENT_VALUE' };
    const nextState = reducer(undefined, action);
    expect(nextState).toMatchSnapshot();
    expect(nextState.value).toStrictEqual(1);
  });

  test('should use the default state when state is null', () => {
    const action: ButtonActions = { type: 'INCREMENT_VALUE' };
    const nextState = reducer(undefined, action);
    expect(nextState).toMatchSnapshot();
    expect(nextState.value).toStrictEqual(1);
  });

  test('should return state for INCREMENT_VALUE', () => {
    const currentState = getCurrentState(4, 'Output', 'Error');
    const action: ButtonActions = { type: 'INCREMENT_VALUE' };
    const nextState = reducer(currentState, action);
    expect(nextState).toMatchSnapshot();
    expect(nextState.value).toStrictEqual(5);
  });

  test('should return state for DECREMENT_VALUE', () => {
    const currentState = getCurrentState(3, 'Output', 'Error');
    const action: ButtonActions = { type: 'DECREMENT_VALUE' };
    const nextState = reducer(currentState, action);
    expect(nextState).toMatchSnapshot();
    expect(nextState.value).toStrictEqual(2);
  });

  test('should return state for SET_VALUE', () => {
    const currentState = getCurrentState(5, 'Output', 'Error');
    const action: ButtonActions = { type: 'SET_VALUE', payload: 1 };
    const nextState = reducer(currentState, action);
    expect(nextState).toMatchSnapshot();
    expect(nextState.value).toStrictEqual(1);
  });

  test('should return state for EXEC_LS', () => {
    const currentState = getCurrentState(5, 'Output', 'Error');
    const action: ButtonActions = { type: 'EXEC_LS' };
    const nextState = reducer(currentState, action);
    expect(nextState).toMatchSnapshot();
    expect(nextState.output).toStrictEqual('');
    expect(nextState.error).toStrictEqual('');
  });

  test('should return state for EXEC_LS_SUCCESS', () => {
    const currentState = getCurrentState(5, '', '');
    const action: ButtonActions = { type: 'EXEC_LS_SUCCESS', payload: 'Success' };
    const nextState = reducer(currentState, action);
    expect(nextState).toMatchSnapshot();
    expect(nextState.output).toStrictEqual('Success');
    expect(nextState.error).toStrictEqual('');
  });

  test('should return state for EXEC_LS_ERROR', () => {
    const currentState = getCurrentState(5, '', '');
    const action: ButtonActions = { type: 'EXEC_LS_ERROR', payload: 'Error' };
    const nextState = reducer(currentState, action);
    expect(nextState).toMatchSnapshot();
    expect(nextState.output).toStrictEqual('');
    expect(nextState.error).toStrictEqual('Error');
  });
});
