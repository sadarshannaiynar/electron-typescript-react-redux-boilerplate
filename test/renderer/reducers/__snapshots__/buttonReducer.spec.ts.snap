// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buttonReducer Tests should return state for DECREMENT_VALUE 1`] = `
Object {
  "error": "Error",
  "output": "Output",
  "value": 2,
}
`;

exports[`buttonReducer Tests should return state for EXEC_LS 1`] = `
Object {
  "error": "",
  "output": "",
  "value": 5,
}
`;

exports[`buttonReducer Tests should return state for EXEC_LS_ERROR 1`] = `
Object {
  "error": "Error",
  "output": "",
  "value": 5,
}
`;

exports[`buttonReducer Tests should return state for EXEC_LS_SUCCESS 1`] = `
Object {
  "error": "",
  "output": "Success",
  "value": 5,
}
`;

exports[`buttonReducer Tests should return state for INCREMENT_VALUE 1`] = `
Object {
  "error": "Error",
  "output": "Output",
  "value": 5,
}
`;

exports[`buttonReducer Tests should return state for SET_VALUE 1`] = `
Object {
  "error": "Error",
  "output": "Output",
  "value": 1,
}
`;

exports[`buttonReducer Tests should return the default state 1`] = `
Object {
  "error": "",
  "output": "",
  "value": 0,
}
`;

exports[`buttonReducer Tests should use the default state when state is null 1`] = `
Object {
  "error": "",
  "output": "",
  "value": 1,
}
`;

exports[`buttonReducer Tests should use the default state when state is undefined 1`] = `
Object {
  "error": "",
  "output": "",
  "value": 1,
}
`;
